### ISSUE TITLE
`infinite_loop` should not trigger when yielding in a gen block

### ISSUE THREAD

**Issue by ronnodas** [rust-lang/rust-clippy#17310]
### Summary

In a `gen` block (or `gen fn`), it is fine to `yield` in an infinite loop, since it is only evaluated lazily as the iterator is driven. I think in a `gen` block, this lint should treat `yield` statements the same as diverging statements like `return` and `break`.

### Lint Name

infinite_loop

### Reproducer

I tried this code ([playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=0a220a40fed1c89173508bc3fabd10f4)):

```rust
#![feature(gen_blocks)]
#![warn(clippy::infinite_loop)]

gen fn repeat() -> u32 {
    loop {
        yield 0;
    }
}
```

I saw this happen:

```
warning: infinite loop detected
 --> src/main.rs:5:5
  |
5 | /     loop {
6 | |         yield 0;
7 | |     }
  | |_____^
  |
  = help: if this is not intended, try adding a `break` or `return` to the loop
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#infinite_loop
```

I expected to see this happen: No lint triggers.

### Version

```text
rustc 1.98.0-nightly (bd08c9e71 2026-06-25)
binary: rustc
commit-hash: bd08c9e71874a81670fe3938dbf85148e42c2b96
commit-date: 2026-06-25
host: x86_64-unknown-linux-gnu
release: 1.98.0-nightly
LLVM version: 22.1.7
```

**Comment by Snxhit** [comment]
@rustbot claim
I will try my hand at this.
